src/main/java/jenkins/branch/NameMangler.java

```diff
@@ -0,0 +1,172 @@
+                    case ".":
+                    case "..":
+                    case "con":
+                    case "prn":
+                    case "aux":
+                    case "nul":
+                    case "com1":
+                    case "com2":
+                    case "com3":
+                    case "com4":
+                    case "com5":
+                    case "com6":
+                    case "com7":
+                    case "com8":
+                    case "com9":
+                    case "lpt1":
+                    case "lpt2":
+                    case "lpt3":
+                    case "lpt4":
+                    case "lpt5":
+                    case "lpt6":
+                    case "lpt7":
+                    case "lpt8":
+                    case "lpt9":
+                        unsafe = true;
+                        break;
+                    default:
+                        if (name.endsWith(".")) {
+                            unsafe = true;
+                        }
+                        break;
+                }
+            }
+            if (!unsafe) {
+                return name;
+            }
+        }
+        StringBuilder buf = new StringBuilder(name.length() + 16);
+        for (char c : name.toCharArray()) {
+            if (isSafe(c)) {
+                buf.append(c);
+            } else if (c == '/' || c == '\\' || c == ' ') {
+                buf.append('_');
+            } else if (c <= 0xff) {
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString(c & 0xff), 2, '0'));
+            } else {
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString(c & 0xff), 2, '0'));
+                buf.append('%');
+                buf.append(StringUtils.leftPad(Integer.toHexString((c & 0xffff) >> 8), 2, '0'));
+            }
+        }
+        // use the digest of the original name
+        String digest;
+        try {
+            MessageDigest sha = MessageDigest.getInstance("SHA-1");
+            byte[] bytes = sha.digest(name.getBytes(StandardCharsets.UTF_8));
+            int bits = 0;
+            int data = 0;
+            StringBuffer dd = new StringBuffer(32);
+            for (byte b : bytes) {
+                while (bits >= 5) {
+                    dd.append(toDigit(data & 0x1f));
+                    bits -= 5;
+                    data = data >> 5;
+                }
+                data = data | ((b & 0xff) << bits);
+                bits += 8;
+            }
+            digest = dd.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("MD5 not installed", e);    // impossible
+        }
+        if (buf.length() <= MAX_SAFE_LENGTH - MIN_HASH_LENGTH - 1) {
+            // we have room to add the min hash
+            buf.append('@');
+            buf.append(StringUtils.right(digest, MIN_HASH_LENGTH));
```

Comment:
Well it's not quite so easy... we prefix with the start of the name... in fact on this branch we prefix with the entire encoded name and then append a hash of the original name... so you only get to play with the characters that have been mapped to `_`... of which the most you can have is 24 of them... and they only have 3 possible values... so that's 3^24 or ~ 38 bits which is 7 characters... so maybe 32 potential collisions for somebody who's branch is called `////////////////////////` if they have a branch with at least one non-flattened character, e.g. `a_______________________` then there are only 3^23 potential names that have the same prefix and that is a space of ~ 6 characters which is the exact amount of hash that we use.

Now for longer names, you are constrained in that the start and end of the name have to match because we insert the hash in the middle for those cases.

Where the name is between 24 and 32 characters long you only get to play with the characters mapping to `_` in the prefix and postfix and the characters in the middle (overwritten by the hash) but at least for those you get to use a wider set (basically `0-9A-Za-z` otherwise it will be `%` encoded and the length could grow and affect how much of the postfix is output) So there is scope there for collisions in a 6 character hash, but you would need to brute-force it *and* have write permission to the repository

After that we start growing the hash length until it is 12 characters of hash